### ISSUE TITLE
Error Loading .pb model file from readNetFromTensorflow

### ISSUE THREAD

**Issue by zy2008cn** [opencv/opencv#14482]
OpenCV => 4.1.0
Operating System / Platform => Windows 64 Bit
Compiler => Visual Studio 2015

**Detailed description**
	String weights = "model_opt.pb";
	String prototxt = "model_opt.pbtxt";	
1)	dnn::Net net = cv::dnn::readNetFromTensorflow(weights);
OpenCV(4.1.0) Error: Unspecified error (More than one input is Const op) in cv::dnn::dnn4_v20190122::`anonymous-namespace'::TFImporter::getConstBlob, file C:\build\master_winpack-build-win64-vc14\opencv\modules\dnn\src\tensorflow\tf_importer.cpp, line 522
 2)   dnn::Net net = cv::dnn::readNetFromTensorflow(weights,prototxt);    
OpenCV(4.1.0) Error: Assertion failed (const_layers.insert(std::make_pair(name, li)).second) in cv::dnn::dnn4_v20190122::`anonymous-namespace'::addConstNodes, file C:\build\master_winpack-build-win64-vc14\opencv\modules\dnn\src\tensorflow\tf_importer.cpp, line 616

**Comment by zy2008cn** [comment]
[models.zip](https://github.com/opencv/opencv/files/3145299/models.zip)

**Comment by zy2008cn** [comment]
Look forward to your favourable reply! @dkurt

**Comment by FadyEssam** [comment]
I imported the "model_opt.pb" and it works fine after the running the TransformGraph script that you provided.

But if you want to import the "model_opt.pbtxt" you should do 2 things:

- add "fold_constants(ignore_errors=True)" to the Transforms to fold the add node of batch normalization

 - remove the assertion from the line opencv/opencv/blob/master/modules/dnn/src/tensorflow/tf_importer.cpp line 663 and recompile opencv

**Comment by dkurt** [comment]
@FadyEssam, removing assertions is not a good idea. It detects unexpected execution so we need to figure out which thing triggered this assertion.

**Comment by FadyEssam** [comment]
I'm just saying if he really needs to import this pbtxt, that's just one way to do it :smiley:

**Comment by FadyEssam** [comment]
@dkurt  because the const layers in the pb file is the exact same in pbtxt file.. and the code extracts const_layers from both files and append them to a map and if a layer name is repeated, it throws an exception.. maybe I don't really understand this code, but I don't see how it wold work otherwise

**Comment by dkurt** [comment]
@FadyEssam, does it mean that using just a `.pb` file we can import and run the model without changing the library's code?

**Comment by FadyEssam** [comment]
@dkurt yes.

**Comment by dkurt** [comment]
@FadyEssam, you may just generate an input blob filled by random values and forward it. Just to check that there is no problem during initialization.

**Comment by FadyEssam** [comment]
@dkurt after running forward I got stuck at the assertion "Unspecified error (Can't create layer "logits/Conv2D" of type "FusedResizeAndPadConv2D")"

Any ideas?

**Comment by dkurt** [comment]
You may try to split it by separate layers: Resize and Conv2D. Or try to disable this fused node (if it's possible) by TransformGraph arguments: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#replacematchingoptypes

**Comment by zy2008cn** [comment]
Thanks for your reply. @dkurt @FadyEssam 
I imported the .pb file again without TransformGraph optimization and  tried forward propagation successfully. But the .pb with TransformGraph optimization ran wrong. After running forward I got stuck at the same  assertion "Unspecified error (Can't create layer "logits/Conv2D" of type "FusedResizeAndPadConv2D" @FadyEssam . The transformGraph tool use fuse_convolutions parameter by default?

**Comment by FadyEssam** [comment]
@zy2008cn I think it's probided as option in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#transform-functions  

I don't know why it would ru  bu default.

**Comment by TMaysGGS** [comment]
Hi, @zy2008cn, @FadyEssam , @dkurt 

I was wondering if you have solved this issue since I met some familiar problems. The error info in forward propagation of mine is: 
Can't create layer "Pnet_prelu3/Neg" of type "Neg". 
But actually in layer "Pnet_prelu1/Neg" & "Pnet_prelu2/Neg" before this layer there are already ops "Neg". 

And this is the error info I got if reading the model without the .pbtxt file. I got the same error info when reading the model with the .pbtxt file. 

My model is simply a Keras implementation of P-Net of MTCNN and I have transfered it into an optimized .pb file. If you @FadyEssam @dkurt are available, could you please check it since this is really a basic Conv Net without even fully connected layers. 

For R-Net and O-Net, since there is a Permute layer to switch the channels of the intermediate input, the 2 nets are also unable to be read. The error is: Parsing error (Only NHWC <-> NCHW permutations are allowed.). 

Hope to hear from you. 
Thanks.

[optimized_Rnet.zip](https://github.com/opencv/opencv/files/3371352/optimized_Rnet.zip)

[optimized_Pnet.zip](https://github.com/opencv/opencv/files/3371320/optimized_Pnet.zip)

**Comment by dkurt** [comment]
@zy2008cn, plase do the following once again over `model_opt.pb`:
```python
import tensorflow as tf
from tensorflow.tools.graph_transforms import TransformGraph

pb_file = 'model_opt.pb'
graph_def = tf.compat.v1.GraphDef()

try:
    with tf.io.gfile.GFile(pb_file, 'rb') as f:
        graph_def.ParseFromString(f.read())
except:
    with tf.gfile.FastGFile(pb_file, 'rb') as f:
        graph_def.ParseFromString(f.read())

graph_def = TransformGraph(graph_def, ['Placeholder'], ['logits/BiasAdd'], ["fold_constants", 'sort_by_execution_order'])

with tf.gfile.FastGFile('model_opt_opt.pb', 'wb') as f:
   f.write(graph_def.SerializeToString())
```

Then you can try a patch from https://github.com/opencv/opencv/pull/16979:
```python
import numpy as np
import cv2 as cv

np.random.seed(21)

import tensorflow as tf

cvNet = cv.dnn.readNet('model_opt_opt.pb')
inp = np.random.randint(0, 255, [1, 3, 80, 80]).astype(np.uint8)
cvNet.setInput(inp.astype(np.float32))
cvOut = cvNet.forward()

print(cvOut.shape)

graph = 'model_opt.pb'
with tf.gfile.FastGFile(graph, 'rb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

with tf.Session() as sess:
    sess.graph.as_default()
    tf.import_graph_def(graph_def, name='')
    tfOut = sess.run(sess.graph.get_tensor_by_name('logits/BiasAdd:0'),
                     feed_dict={'Placeholder:0': inp.transpose(0, 2, 3, 1)})

print(np.max(np.abs(tfOut.transpose(0, 3, 1, 2) - cvOut)))
```

max diff: `0.00049591064`

**Comment by dkurt** [comment]
@TMaysGGS, thanks for this note, please try to use the changes from https://github.com/opencv/opencv/pull/16983 and the following steps:

for RNet:
1. Do extra constants fusion
```python
import tensorflow as tf
from tensorflow.tools.graph_transforms import TransformGraph

pb_file = 'optRnet.pb'
graph_def = tf.compat.v1.GraphDef()

try:
    with tf.io.gfile.GFile(pb_file, 'rb') as f:
        graph_def.ParseFromString(f.read())
except:
    with tf.gfile.FastGFile(pb_file, 'rb') as f:
        graph_def.ParseFromString(f.read())

graph_def = TransformGraph(graph_def, ['Rnet_input_1'], ['Rnet_bbox_regressor_1/BiasAdd'], ["fold_constants", 'sort_by_execution_order'])

 with tf.gfile.FastGFile('optRnet_2.pb', 'wb') as f:
    f.write(graph_def.SerializeToString())
```
2. Ru with OpenCV (difference between OpenCV and TensorFlow is `7.4505806e-08`):
```python
import numpy as np
import cv2 as cv

np.random.seed(21)

import tensorflow as tf

cvNet = cv.dnn.readNet('optRnet_2.pb')
inp = np.random.standard_normal([1, 3, 24, 24]).astype(np.float32)
cvNet.setInput(inp)
cvOut = cvNet.forward()

graph = 'optRnet.pb'
with tf.gfile.FastGFile(graph, 'rb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

with tf.Session() as sess:
    sess.graph.as_default()
    tf.import_graph_def(graph_def, name='')
    tfOut = sess.run(sess.graph.get_tensor_by_name('Rnet_bbox_regressor_1/BiasAdd:0'),
                     feed_dict={'Rnet_input_1:0': inp.transpose(0, 2, 3, 1)})

print(np.max(np.abs(tfOut - cvOut)))
```

---

For PNet you may run it as is:
```python
import numpy as np
import cv2 as cv

np.random.seed(21)

import tensorflow as tf

cvNet = cv.dnn.readNet('optPnet.pb')
inp = np.random.standard_normal([1, 3, 24, 24]).astype(np.float32)
cvNet.setInput(inp)
cvOut = cvNet.forward(['Pnet_classifier_conv/truediv', 'Pnet_bbox_regressor_conv/convolution'])

graph = 'optPnet.pb'
with tf.gfile.FastGFile(graph, 'rb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

with tf.Session() as sess:
    sess.graph.as_default()
    tf.import_graph_def(graph_def, name='')
    tfOut = sess.run([sess.graph.get_tensor_by_name('Pnet_classifier_conv/truediv:0'),
                      sess.graph.get_tensor_by_name('Pnet_bbox_regressor_conv/BiasAdd:0')],
                     feed_dict={'Pnet_input:0': inp.transpose(0, 2, 3, 1)})

print(tfOut[0].shape, tfOut[1].shape)
print(cvOut[0].shape, cvOut[1].shape)
print(np.max(np.abs(tfOut[0].transpose(0, 3, 1, 2) - cvOut[0])))
print(np.max(np.abs(tfOut[1].transpose(0, 3, 1, 2) - cvOut[1])))
```
output:
```
1.7881393e-07
5.9604645e-07
```

**Comment by TMaysGGS** [comment]
> @TMaysGGS, thanks for this note, please try to use the changes from #16983 and the following steps:
> 
> for RNet:
> 
> 1. Do extra constants fusion
> 
> ```python
> import tensorflow as tf
> from tensorflow.tools.graph_transforms import TransformGraph
> 
> pb_file = 'optRnet.pb'
> graph_def = tf.compat.v1.GraphDef()
> 
> try:
>     with tf.io.gfile.GFile(pb_file, 'rb') as f:
>         graph_def.ParseFromString(f.read())
> except:
>     with tf.gfile.FastGFile(pb_file, 'rb') as f:
>         graph_def.ParseFromString(f.read())
> 
> graph_def = TransformGraph(graph_def, ['Rnet_input_1'], ['Rnet_bbox_regressor_1/BiasAdd'], ["fold_constants", 'sort_by_execution_order'])
> 
>  with tf.gfile.FastGFile('optRnet_2.pb', 'wb') as f:
>     f.write(graph_def.SerializeToString())
> ```
> 
> 1. Ru with OpenCV (difference between OpenCV and TensorFlow is `7.4505806e-08`):
> 
> ```python
> import numpy as np
> import cv2 as cv
> 
> np.random.seed(21)
> 
> import tensorflow as tf
> 
> cvNet = cv.dnn.readNet('optRnet_2.pb')
> inp = np.random.standard_normal([1, 3, 24, 24]).astype(np.float32)
> cvNet.setInput(inp)
> cvOut = cvNet.forward()
> 
> graph = 'optRnet.pb'
> with tf.gfile.FastGFile(graph, 'rb') as f:
>     graph_def = tf.GraphDef()
>     graph_def.ParseFromString(f.read())
> 
> with tf.Session() as sess:
>     sess.graph.as_default()
>     tf.import_graph_def(graph_def, name='')
>     tfOut = sess.run(sess.graph.get_tensor_by_name('Rnet_bbox_regressor_1/BiasAdd:0'),
>                      feed_dict={'Rnet_input_1:0': inp.transpose(0, 2, 3, 1)})
> 
> print(np.max(np.abs(tfOut - cvOut)))
> ```
> 
> For PNet you may run it as is:
> 
> ```python
> import numpy as np
> import cv2 as cv
> 
> np.random.seed(21)
> 
> import tensorflow as tf
> 
> cvNet = cv.dnn.readNet('optPnet.pb')
> inp = np.random.standard_normal([1, 3, 24, 24]).astype(np.float32)
> cvNet.setInput(inp)
> cvOut = cvNet.forward(['Pnet_classifier_conv/truediv', 'Pnet_bbox_regressor_conv/convolution'])
> 
> graph = 'optPnet.pb'
> with tf.gfile.FastGFile(graph, 'rb') as f:
>     graph_def = tf.GraphDef()
>     graph_def.ParseFromString(f.read())
> 
> with tf.Session() as sess:
>     sess.graph.as_default()
>     tf.import_graph_def(graph_def, name='')
>     tfOut = sess.run([sess.graph.get_tensor_by_name('Pnet_classifier_conv/truediv:0'),
>                       sess.graph.get_tensor_by_name('Pnet_bbox_regressor_conv/BiasAdd:0')],
>                      feed_dict={'Pnet_input:0': inp.transpose(0, 2, 3, 1)})
> 
> print(tfOut[0].shape, tfOut[1].shape)
> print(cvOut[0].shape, cvOut[1].shape)
> print(np.max(np.abs(tfOut[0].transpose(0, 3, 1, 2) - cvOut[0])))
> print(np.max(np.abs(tfOut[1].transpose(0, 3, 1, 2) - cvOut[1])))
> ```
> 
> output:
> 
> ```
> 1.7881393e-07
> 5.9604645e-07
> ```

Hi @dkurt , 
Thanks very much for your help. 

After that problem happened, I have to transfer the model to caffemodel and then loaded it by using cv2.dnn.readNetFromCaffe API. But your solution is more straightforward to this problem. 

I wonder if your modification to the layer support for TensorFlow will be added into OpenCV in the future like 4.3 or above, or I should build one myself for the two new layers since they are frequently used.

**Comment by dkurt** [comment]
@TMaysGGS, OpenCV 4.3.0 have been already released. There is a workaround for existing releases using `.pbtxt` file (read https://github.com/opencv/opencv/wiki/TensorFlow-text-graphs). You need to split FusedResizeAndPadConv2D onto two nodes. And probably other issues with transposing, PReLU replacement. If it's critical we can try to propose a `.pbtxt` file for you.

**Comment by TMaysGGS** [comment]
> @TMaysGGS, OpenCV 4.3.0 have been already released. There is a workaround for existing releases using `.pbtxt` file (read https://github.com/opencv/opencv/wiki/TensorFlow-text-graphs). You need to split FusedResizeAndPadConv2D onto two nodes. And probably other issues with transposing, PReLU replacement. If it's critical we can try to propose a `.pbtxt` file for you.

Thank you @dkurt , I'll try to follow your advice and see how far I can go.
